spec/puppet/resource_api_spec.rb

```diff
@@ -404,56 +414,124 @@ def set(_context, changes)
         end
       end
     end
+    let(:log_sink) { [] }
 
     before(:each) do
       stub_const('Puppet::Provider::Canonicalizer', Module.new)
       stub_const('Puppet::Provider::Canonicalizer::Canonicalizer', provider_class)
+
+      Puppet::Util::Log.newdestination(Puppet::Test::LogCollector.new(log_sink))
+    end
+
+    after(:each) do
+      log_sink.clear
     end
 
     it { expect { described_class.register_type(definition) }.not_to raise_error }
 
     describe 'the registered type' do
       subject(:type) { Puppet::Type.type(:canonicalizer) }
 
+      let(:first_test_string_set) { 'foo' }
+      let(:first_test_string_return) { 'foo' }
+      let(:second_test_string_set) { 'wibble' }
+      let(:second_test_string_return) { 'wibble' }
+      let(:instance) { type.new(name: 'somename', test_string: first_test_string_return) }
+      let(:other_instance) { type.new(name: 'other', test_string: second_test_string_return) }
+      let(:log_message) do
+        <<MESSAGE.strip
+canonicalizer[somename]#get has not provided canonicalized values.
+Returned values:       {:name=>"somename", :test_string=>"foo"}
+Canonicalized values:  {:name=>"somename", :test_string=>"canonfoo"}
+MESSAGE
+      end
```

Comment:
I'm not understanding what the different values here are used/useful for. Let's work on this bit after lunch?